wasm2c/wasm-rt-impl-mem.c

```diff
@@ -0,0 +1,157 @@
+// memory operations on different memory tyeps --- wasm_rt_memory_t and
+// wasm_rt_memory_shared_t depending on the value of WASM_RT_SHARED_MEM_OPS.
+
+#include "wasm-rt-impl.h"
+
+#ifndef WASM_RT_SHARED_MEM_OPS
+#error "Expected value for WASM_RT_SHARED_MEM_OPS"
+#endif
+
+#if WASM_RT_SHARED_MEM_OPS == 0
+// Memory operations on wasm_rt_memory_t
+#define MEMORY_TYPE_NAME wasm_rt_memory_t
+#define MEMORY_DATA_TYPE uint8_t*
+#define MEMORY_API_NAME(name) name
+#else
+// Memory operations on wasm_rt_shared_memory_t
+#define MEMORY_TYPE_NAME wasm_rt_shared_memory_t
+#define MEMORY_DATA_TYPE _Atomic volatile uint8_t*
+#define MEMORY_API_NAME(name) name##_shared
+#endif
+
+#define PAGE_SIZE 65536
+
+#if WASM_RT_USE_MMAP
+
+static uint64_t MEMORY_API_NAME(get_memory_allocation_size_for_mmap)(
+    MEMORY_TYPE_NAME* memory) {
+  assert(!memory->is64 &&
+         "memory64 is not yet compatible with WASM_RT_USE_MMAP");
+#if WASM_RT_MEMCHECK_GUARD_PAGES
+  /* Reserve 8GiB. */
+  const uint64_t max_size = 0x200000000ul;
+  return max_size;
+#else
+  if (memory->max_pages != 0) {
+    const uint64_t max_size = memory->max_pages * PAGE_SIZE;
+    return max_size;
+  }
+
+  /* Reserve 4GiB. */
+  const uint64_t max_size = 0x100000000ul;
+  return max_size;
+#endif
+}
+
+#endif
+
+void MEMORY_API_NAME(wasm_rt_allocate_memory)(MEMORY_TYPE_NAME* memory,
+                                              uint64_t initial_pages,
+                                              uint64_t max_pages,
+                                              bool is64) {
+  uint64_t byte_length = initial_pages * PAGE_SIZE;
+  memory->size = byte_length;
+  memory->pages = initial_pages;
+  memory->max_pages = max_pages;
+  memory->is64 = is64;
+
+#if WASM_RT_USE_MMAP
+  const uint64_t mmap_size =
+      MEMORY_API_NAME(get_memory_allocation_size_for_mmap)(memory);
+  void* addr = os_mmap(mmap_size);
+  if (!addr) {
+    os_print_last_error("os_mmap failed.");
+    abort();
+  }
+  int ret = os_mprotect(addr, byte_length);
+  if (ret != 0) {
+    os_print_last_error("os_mprotect failed.");
+    abort();
+  }
+  memory->data = addr;
+#else
+  memory->data = calloc(byte_length, 1);
+#endif
+}
+
+static uint64_t MEMORY_API_NAME(grow_memory_impl)(MEMORY_TYPE_NAME* memory,
+                                                  uint64_t delta) {
+  uint64_t old_pages = memory->pages;
+  uint64_t new_pages = memory->pages + delta;
+  if (new_pages == 0) {
+    return 0;
+  }
+  if (new_pages < old_pages || new_pages > memory->max_pages) {
+    return (uint64_t)-1;
+  }
+  uint64_t old_size = old_pages * PAGE_SIZE;
+  uint64_t new_size = new_pages * PAGE_SIZE;
+  uint64_t delta_size = delta * PAGE_SIZE;
+#if WASM_RT_USE_MMAP
+  MEMORY_DATA_TYPE new_data = memory->data;
+  int ret = os_mprotect((void*)(new_data + old_size), delta_size);
+  if (ret != 0) {
+    return (uint64_t)-1;
+  }
+#else
+  MEMORY_DATA_TYPE new_data = realloc((void*)memory->data, new_size);
```

Comment:
realloc is potentially unsound for shared memories, if multiple threads are active, the allocation could be moved while other threads are accessing it.

Reply:
yup! I was initially planning on making these changes in a new PR along with wait/notify runtime support. However, on re-thinking it may make sense to ensure this PR is self contained. Will fix!